File: js-boj/src/9466_텀_프로젝트.js
const rl = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

let t;
let cur = 1;
const nInput = [];
const pickInput = [];

rl.on("line", (input) => {
  if (!t) t = +input;
  else if (cur <= t) {
    if (nInput.length < cur) {
      nInput.push(+input);
    } else if (pickInput.length < cur) {
      pickInput.push(input.split(" ").map((i) => +i));

      if (cur < t) cur++;
      else rl.close();
    }
  }
}).on("close", () => {
  for (let i = 0; i < t; i++) {
    console.log(sol(i));
  }
});

const sol = (i) => {
  const n = nInput[i];
  const picks = [-1, ...pickInput[i]];
  const isVisited = Array(n + 1).fill(false);
  const isDone = Array(n + 1).fill(false);
  let ans = n;

  const dfs = (now) => {
    isVisited[now] = true;
    const next = picks[now];
    
    if(!isVisited[next]) {
      dfs(next);
    }
    else if(!isDone[next]) {
      for(let i = next; i !== now; i = picks[i]) {
        ans--;
      }
      ans--;
    }
    isDone[now] = true;
  }

  for(let i = 1; i <= n; i ++) {
    if(!isVisited[i]) {
      dfs(i);
    }
  }

  return ans;
}

File: algorithm_from_2021/src/main/kotlin/self_boj/9466_텀_프로젝트.kt
package self_boj

private fun main() {
    val t = readLine()!!.toInt()

    repeat(t) {
        val n = readLine()!!.toInt()
        val picks = readLine()!!.split(" ").map { it.toInt() }
        val isVisited = BooleanArray(n + 1)
        val isDone = BooleanArray(n + 1)
        var ans = n

        fun dfs(now: Int) {
            isVisited[now] = true
            val next = picks[now - 1]

            if(!isVisited[next]) {
                dfs(next)
            }
            else if(!isDone[next]) {
                var i = next
                while(i != now) {
                    ans--
                    i = picks[i - 1]
                }
                ans--
            }
            isDone[now] = true
        }

        for(i in 1..n) {
            if(!isVisited[i]) {
                dfs(i)
            }
        }

        println(ans)
    }
}